armi/bookkeeping/db/database3.py

```diff
@@ -1585,6 +1771,15 @@ def packSpecialData(
 
         return data, attrs
 
+    if isinstance(arrayData, JaggedArray):
+        attrs["jagged"] = True
+        attrs["offsets"] = arrayData.offsets
+        attrs["shapes"] = arrayData.shapes
+        attrs["noneLocations"] = arrayData.nones
+        if len(arrayData.nones > 0):
+            data = replaceNonesWithNonsense(data, paramName, arrayData.nones)
+        return data, attrs
+
     # conform non-numpy arrays to numpy
```

Comment:
We might want to consider `assert`ing that the data passed into `packSpecialData` is either a `numpy.ndarray` or `JaggedArray` instance. If some kind of `list`-of-`tuple`, `tuple`-of-`list`s, etc. is passed in, it should probably be converted to either a `JaggedArray` (if jagged) or `numpy.ndarray` (if not jagged) and then passed in.

Reply:
> So, your implied thesis here is that the place where ARMI (and its ecosystem) mostly ends up `JaggedArray`s is when writing data to the DB?

Sorry, I was out for a couple weeks and just now circling back to this.

Yes, I don't know of anywhere ARMI creates a jagged array and tries to pass it into `numpy` other than when it's writing to a database. It was a shortcut for packing data; we would just do `a = numpy.array(jaggedData)`, and then `a.flatten()` to get a 1D representation of it.

The jagged data mainly comes from pin-level data in blocks -- for example, the parameter `b.p.linPowByPin`. During the course of a simulation, each block with pins has this parameter stored as a 1D array. When it is time to store this data in the database, ARMI puts all of the `linPowByPin` data into a single array. Since different blocks can have different number of pins, this is a jagged array.